### ISSUE TITLE
[Feature] Provide EntryComponents in Root Injector

### ISSUE THREAD

**Issue by lucasbasquerotto** [angular/angular#23819]
## I'm submitting a...

<pre><code>[x] Feature request</code></pre>

## Current behavior

I need to add components in the entryComponents array of a NgModule so I can use it with `ComponentFactoryResolver`:

`this.componentFactoryResolver.resolveComponentFactory<MyComponent>(componentType);`

## Expected behavior

Adds a `providesIn` property in the `@Component()` annotation so that it behaves just like `@Injectable()` that, in the new Angular 6 version, doesn't need to be declared in the `providers` array, making it much better both from a performance perspective as well as from a maintenance perspective. In the case of EntryComponents, there wouldn't be a performance gain (because a new instance will always be created and you don't need to declare them in app module), but there is the maintenance gain.

Then it would be possible to define a component as `@Component( ... providesIn: 'root' )` and use the component as an entry component without defining it in a module. The component code would be included in the modules that use it.

I don't know how that is done internally tough, but I think there should be a way to implement it similar to how it handles Injectables. 

Maybe the `providesIn` is unnecessary because you don't inject them as services, but creates a new instance always, so the only thing needed was to remove the need to include in the `entryComponents` array.

## What is the motivation / use case for changing the behavior?

The motivation is basically the same as the one to allow services root scoped without having to import them in `AppModule` (or any module).

Here is an issue I created about providers (services), then I saw in an answer to my request about the new feature about providers (great feature, Angular really needed it for medium and large apps):

https://github.com/angular/angular/issues/23770#issuecomment-387439265

Although the issue with providers was worse because there is a high probability of wanting to make them as singletons, while with components a new instance will always be created, it makes easy to maintain without including it in the modules that need it (after all, it will be created in an imperative way, and the reference will be known).

## Environment

<pre><code>Angular version: 6</code></pre>

**Comment by trotyl** [comment]
> making it much better both from a performance perspective

It won't, components are already tree-shakable now.

> In the case of EntryComponents, there wouldn't be a performance gain

There will be a performance slowdown, `entryComponents` mean it require generating extra code than normal components, your making it indistinguishable will results to much extra size.

**Comment by lucasbasquerotto** [comment]
@trotyl About the first quote I was talking about providers/services, then I complemented with the EntryComponents (your 2nd quote).

About the slowdown I don't think so, the component will be only included in the modules that have some component that use/import them explicitly (if they aren't used, they wouldn't be included in the module). If they are used only in lazy loaded modules, they would only be included in them and not in the initial load.

So the `providesIn` (or similar approach) would actually be better, because the way it is now (having to import in a module) it may be included even if you don't use it (the code would go to the bundle anyway), or if you are using it in some component but then stopped using and forgot to remove from the module.

The performance gain would be only in such cases tough. If it is included only where it is used, there would be no performance gain (my 2nd quote), but the maintenance would be easier.

> It won't, components are already tree-shakable now.

Is this true even if the component is included in entryComponents and not used in the module? If that is so, there wouldn't be a performance gain in any case, but I don't see why it would have a performance slowdown, having in mind what I said above (if the component is not used in the module, it won't be included, if extra code is generated, it would be generated anyway (even with entryComponents) because some component in the module is importing it).

I don't know how it is done internally tough, but I hope that the same approach that was done with providers can be done with components.

**Comment by trotyl** [comment]
> About the slowdown I don't think so, the component will be only included in the modules that have some component that use/import them explicitly

Not true for entry components, entry components are not tree-shakable in any way. What you imported is the component type, but the `ComponentFactory` type must be registered to the runtime so you can retrieve that via `ComponentFactoryResolver`, namely imported by others codes. There would be significant if you treat every component as entry component.

So you still need to config it like:

```typescript
@Component({
  entryComponent: true
})
class MyComp {}
```

> So the providesIn (or similar approach) would actually be better, because the way it is now (having to import in a module) it may be included even if you don't use it (the code would go to the bundle anyway)

It's only true for `providers`, not for `declarations`. `providers` would be compiled to something in the `NgModule`, but `declarations` just work as the compiler config, itself results nothing in `NgModule`, but affect compilation process in other place. 

`declarations` are already tree-shakable in current way of working. (But I agree about the ergonomics part)

**Comment by lucasbasquerotto** [comment]
@trotyl ATM, I include all my components as `entryComponents`, but that is because I consider that a better design (otherwise the component should "know" who are its consumers, what I consider a bad design). 

I don't know what angular does internally, tough, so I don't know what performance degradation that could lead to (I expect that the code to define a component as an EntryComponent is small compared to the component itself).

From what you say, that might not be the case. I will take a look to see if removing the components from the `entryComponents` array change performance in a significative way.

Just to be clear, **what I want is the following**, similar to what I would expect from providers:

If the lazy loaded modules `LazyLoadedModuleA` and `LazyLoadedModuleB` have the components `ComponentA` and `ComponentB`, respectively, and both components use dynamically (via `ComponentFactoryResolver`) the component `ComponentC`, and `ComponentC` has `providedIn: 'root'` (or `entryComponent: true`, or anything similar), then:

1. In the build, the `ComponentC` (and its factory) code will be in the modules that have components that use/import it (if only lazy loaded modules use it, it won't be included in the initial load).
2. If the chunks of `LazyLoadedModuleA` and `LazyLoadedModuleB` are `1.js` and `2.js`, respectivelly, both chunks would have `ComponentC` and its factory, but `main.js` and all other chunks wouldn't.
3. `LazyLoadedModuleA` is loaded and `ComponentA ` can create `ComponentC` dynamically.
4. `LazyLoadedModuleB` is loaded and `ComponentB ` can create `ComponentC` dynamically.

All of this without having to include `ComponentC` it in the `entryComponents` of any module.

So this feature is about including both the components **and their factories** only in the modules that includes them. I hope this is possible.

I achieve that now, but I need to import a module that has the `entryComponents` defined and will only receive an error in runtime if I forget.

**Comment by mlc-mlapis** [comment]
@lucasbasquerotto ... some comments:

ad 1) ... it is supposed that a reference to `ComponentC` is directly used just in that lazy loaded module (as a tag element in a template, related to a route). Another thing is a dynamic / programmatic way, where no direct reference exists so no code is aware of it. And that is why it has to be included in `entryComponents` section.

ad 2) ... looks strange to duplicate `ComponentC` in modules and it's also not possible at all actually.

ad 3) ... how the compiler should know that your code wants to create `ComponentC` dynamically? Do you think that it is possible to analyse the code somehow and find out which component it should be?

ad 4) ... the same as 3).

The reason why to add a component to `entryComponents` is only one ... compiler includes only those component factories that are used in templates or which are related to routes into a module.

**Comment by trotyl** [comment]
@lucasbasquerotto  OK, then talking about only `entryComponents` now.

> In the build, the ComponentC (and its factory) code will be in the modules that have components that use/import it (if only lazy loaded modules use it, it won't be included in the initial load).

It's already the case today, just how code-splitting works.

But that won't change the fact you still need to use/import the NgModule declaring it, so there cannot be a `root` level component as its factory will only be shipped by NgModule.

The only possible way to achieve it (without importing its NgModule) now is to import the generated factory manually rather than import the component, like:

```typescript
import { CComponentNgFactory } from './c.component.ngfactory'
```

> If the chunks of LazyLoadedModuleA and LazyLoadedModuleB are 1.js and 2.js, respectivelly, both chunks would have ComponentC and its factory, but main.js and all other chunks wouldn't.

It's nothing related to Angular but your build tool, for example whether something is shared would be controlled by [CommonsChunkPlugin](https://webpack.js.org/plugins/commons-chunk-plugin/) or [SplitChunksPlugin](https://webpack.js.org/plugins/split-chunks-plugin/).

> LazyLoadedModuleA is loaded and ComponentA can create ComponentC dynamically.
> LazyLoadedModuleB is loaded and ComponentB can create ComponentC dynamically.

That's what https://github.com/angular/angular/issues/14324 for.

**Comment by lucasbasquerotto** [comment]
@mlc-mlapis some explanations:

> it is supposed that a reference to ComponentC is directly used just in that lazy loaded module (as a tag element in a template or in relation to a route). 

If it's used in the template it would be using the `declarations` array as it already is, I don't see `entryComponents` allowing to include in the template.

When using in a route it seems angular [already includes](https://angular.io/guide/entry-components#a-bootstrapped-entry-component) the component in the `entryComponents`, I don't think it would cause some conflicts.

> Another thing is a dynamic / programmatic way, where no direct reference exists so no code is aware of it. 

If there is no direct reference in that module, it would be in some other module, otherwise the component reference couldn't be imported. The factory would be registered in the resolver where the component is imported (not in the one where it is created, unless both are the same).

I did some test with providers. I downloaded the example in https://angular.io/guide/router that is using angular6, then created the following service that will receive a **service class** from somewhere:

```
import { Injectable } from '@angular/core';

@Injectable({ providedIn: 'root' })
export class TestService {
	
	private service: any;

	public setService(service: any) {
		this.service = service;		
	}

	public getService() {
		return this.service;
	}
	
}

```

and also the service:

```
import { Injectable } from '@angular/core';

@Injectable({ providedIn: 'root' })
export class ProfileService {
	
	public test() {
		return 'ProfileService->test';		
	}
	
}

```

Then in `CrisisService` that is inside of `CrisisCenterModule` (lazy loaded) I included:

```
constructor(testService: TestService) {
    testService.setService(ProfileService);
}
```

And in `ManageHeroesComponent` that is inside of `AdminModule` (lazy loaded) I included:

```
@Component(...)
export class ManageHeroesComponent {
	constructor(injector: Injector, testService: TestService) {
		let service: { test: () => string } = testService.getService();

		if (service) {
			service = injector.get(service);
			console.log('service', service);
			console.log('method', service.test());
		}
	}
}
```

The `ProfileService` is included only in the chunk of the `CrisisCenterModule` (I looked at the js code), but  by sending its reference to `TestService` (that is agnostic of the what is passed to it) I was able to inject the service inside `ManageHeroesComponent` even tough its module `AdminModule` don't have any direct nor indirect reference to `ProfileService` (the reference was only defined in runtime).

The logs:

```
service [ProfileService instance]
method ProfileService->test
```

I assume that when the code is loaded the factory is defined to the injector to be able to be used by other modules, even if these modules themselves don't have any reference (during compilation) to the service.

What I want to the `entryComponents` is something along these lines, if possible (defining its factory in the resolver when the code is loaded, so when the reference is passed around even if some other module calls the resolver it will be able to instantiate it).

> looks strange to duplicate ComponentC in modules and it's also not possible at all actually. Also it seems to be senseless.

That is possible using `--no-common-chunk` and [there are use cases for it](https://github.com/angular/angular-cli/issues/7021), especially if you have lots of modules with dependencies to the common chunk and the common chunk is huge. Of course, there are pros and cons to this approach (but this doesn't conflict with this proposal, the difference is that the component would be in the common chunk).

> how the compiler should know that your code wants to create ComponentC dynamically

It wouldn't know, it would register the factory anyway in the resolver based on where it is imported, when the chunk that has it (be it a common chunk or not) is loaded, but only if the component is imported and used somewhere, and in most cases, maybe all, the component reference would be used to instantiate it somewhere (I don't think you would be importing it just to log it in your prod build :)). 

My only concern is in the case you only use it in the declarations array (to use in the html directly), but not as an entry component, because I don't know how much more code boilerplate will be generated by making the component as an entry component.

In any case I should make it clear that **my main reason for creating this feature request is to avoid runtime errors caused when trying to create a component dynamically using the factory resolver**.

I don't have problems when components  are used in the template (they are declared in the `declarations` array) because I have compile time errors when running with AOT (not with JIT, but that is fine).

If the factory is registered in the resolver (I don't know if this is exactly what happens behind the scenes) when the component code is loaded, so that it can be created anywhere, **it would solve the problem and as a plus it would have a productivity gain** (because I don't have to import its module anywhere whan I want to use it as an entry component).

I have a concern about the module that has the component in the declarations tough (how the component dependencies included in the module would work in this case). Maybe enforcing that a component with `entryComponent: true` must be included in the entryComponents of a module, and the module would be always bundled together with the component (even when only the component is imported, not sure how this would be achieved tough).

Seeing your response and of @trotyl I'm not too optimistic that this will be implemented (and I don't know how complex it would be, because I'm just seeing it from the outside, I think both of you are more experienced about how angular works internally), but I consider this feature (or some equivalent one) very important in regards to make the code easier to maintain (with less unexpected runtime errors).
